pgml-extension/src/config.rs

```diff
@@ -1,16 +1,70 @@
+use lazy_static::lazy_static;
+use pgrx::{GucContext, GucFlags, GucRegistry, GucSetting};
 use std::ffi::CStr;
 
 #[cfg(any(test, feature = "pg_test"))]
 use pgrx::{pg_schema, pg_test};
-use pgrx_pg_sys::AsPgCStr;
 
-pub fn get_config(name: &str) -> Option<String> {
-    // SAFETY: name is not null because it is a Rust reference.
-    let ptr = unsafe { pgrx_pg_sys::GetConfigOption(name.as_pg_cstr(), true, false) };
-    (!ptr.is_null()).then(move || {
-        // SAFETY: assuming pgrx_pg_sys is providing a valid, null terminated pointer.
-        unsafe { CStr::from_ptr(ptr) }.to_string_lossy().to_string()
-    })
+lazy_static! {
+    pub static ref PGML_VENV: (&'static str, GucSetting<Option<&'static CStr>>) =
+        ("pgml.venv", GucSetting::<Option<&'static CStr>>::new(None));
+    pub static ref PGML_HF_WHITELIST: (&'static str, GucSetting<Option<&'static CStr>>) = (
+        "pgml.huggingface_whitelist",
+        GucSetting::<Option<&'static CStr>>::new(None),
+    );
+    pub static ref PGML_HF_TRUST_REMOTE_CODE: (&'static str, GucSetting<bool>) =
+        ("pgml.huggingface_trust_remote_code", GucSetting::<bool>::new(false));
+    pub static ref PGML_HF_TRUST_WHITELIST: (&'static str, GucSetting<Option<&'static CStr>>) = (
+        "pgml.huggingface_trust_remote_code_whitelist",
+        GucSetting::<Option<&'static CStr>>::new(None),
+    );
+    pub static ref PGML_OMP_NUM_THREADS: (&'static str, GucSetting<i32>) =
+        ("pgml.omp_num_threads", GucSetting::<i32>::new(-1));
+}
+
+pub fn initialize_server_params() {
+    GucRegistry::define_string_guc(
+        PGML_VENV.0,
+        "Python's virtual environment path",
+        "",
+        &PGML_VENV.1,
+        GucContext::Userset,
+        GucFlags::default(),
+    );
+    GucRegistry::define_string_guc(
+        PGML_HF_WHITELIST.0,
+        "Models allowed to be downloaded from huggingface",
+        "",
+        &PGML_HF_WHITELIST.1,
+        GucContext::Userset,
+        GucFlags::default(),
+    );
+    GucRegistry::define_bool_guc(
+        PGML_HF_TRUST_REMOTE_CODE.0,
+        "Whether model can execute remote codes",
+        "",
+        &PGML_HF_TRUST_REMOTE_CODE.1,
+        GucContext::Userset,
+        GucFlags::default(),
+    );
+    GucRegistry::define_string_guc(
+        PGML_HF_TRUST_WHITELIST.0,
+        "Models allowed to execute remote codes when pgml.hugging_face_trust_remote_code = 'on'",
+        "",
+        &PGML_HF_TRUST_WHITELIST.1,
+        GucContext::Userset,
+        GucFlags::default(),
+    );
+    GucRegistry::define_int_guc(
```

Comment:
If we use the `once_cell::sync::Lazy`, what do you think about having a `Lazy` value for each of these config values in the `config` module? Something like this:

```rust
static OMP_NUM_THREADS: Lazy<i32> = Lazy::new(|| {
    let guc = GucSetting::<i32>::new(-1);
    GucRegistry::define_int_guc(
        "pgml.omp_num_threads",
         "",
         "Specifies the number of threads used by default of underlying OpenMP library. Only positive integers are valid.",
         &guc ,
         -1,
         i32::MAX,
         GucContext::Backend,
         GucFlags::default()
    );
    // ensure value is positive
    guc.get().max(0)
});

```

Reply:
Thanks! It works but I want to move the definition of GUC values to the `_PG_init()` function, since it's the place that most extensions define their GUC values.

Reply:
That makes sense that the `Lazy` values need to be initialized from `_PG_init()`. If you like how self-contained the example above is, but still want the `Lazy` values initialized from `_PG_init()` you could do the following inside `_PG_init()` or your `initialize_server_params` function:

```rust
Lazy::force(&OMP_NUM_THREADS);
```

which [forces the evaluation of the Lazy value](https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html#method.force). I'll let you decide.

Reply:
I realize that pgrx has already wrap global variable into `GucSetting<>`. We don't need any additional wrapper around it. So, I just removed `lazy_static` and `Lazy` things.